boot/systems_test.go

```diff
@@ -0,0 +1,535 @@
+	mtbl.BootChainList = []bootloader.BootFile{
+		bootloader.NewBootFile("", "asset", bootloader.RoleRunMode),
+		runKernelBf,
+	}
+	mtbl.RecoveryBootChainList = []bootloader.BootFile{
+		bootloader.NewBootFile("", "asset", bootloader.RoleRecovery),
+		recoveryKernelBf,
+	}
+	bootloader.Force(mtbl)
+	return mtbl
+}
+
+func (s *systemsSuite) SetUpTest(c *C) {
+	s.baseBootenvSuite.SetUpTest(c)
+	c.Assert(os.MkdirAll(boot.InitramfsUbuntuBootDir, 0755), IsNil)
+	c.Assert(os.MkdirAll(boot.InitramfsUbuntuSeedDir, 0755), IsNil)
+
+	restore := boot.MockSecbootResealKeys(func(params *secboot.ResealKeysParams) error { return nil })
+	s.AddCleanup(restore)
+
+	s.uc20dev = boottest.MockUC20Device("", nil)
+
+	// run kernel
+	s.runKernelBf = bootloader.NewBootFile("/var/lib/snapd/snap/pc-kernel_500.snap",
+		"kernel.efi", bootloader.RoleRunMode)
+	// seed (recovery) kernel
+	s.recoveryKernelBf = bootloader.NewBootFile("/var/lib/snapd/seed/snaps/pc-kernel_1.snap",
+		"kernel.efi", bootloader.RoleRecovery)
+}
+
+func (s *systemsSuite) TestSetTryRecoverySystemEncrypted(c *C) {
+	mockAssetsCache(c, s.rootdir, "trusted", []string{
+		"asset-asset-hash-1",
+	})
+
+	mtbl := s.mockTrustedBootloaderWithAssetAndChains(c, s.runKernelBf, s.recoveryKernelBf)
+	bootloader.Force(mtbl)
+	defer bootloader.Force(nil)
+
+	// system is encrypted
+	s.stampSealedKeys(c, s.rootdir)
+
+	modeenv := &boot.Modeenv{
+		Mode: "run",
+		// keep this comment to make old gofmt happy
+		CurrentRecoverySystems: []string{"20200825"},
+		CurrentKernels:         []string{},
+		CurrentTrustedRecoveryBootAssets: boot.BootAssetsMap{
+			"asset": []string{"asset-hash-1"},
+		},
+
+		CurrentTrustedBootAssets: boot.BootAssetsMap{
+			"asset": []string{"asset-hash-1"},
+		},
+	}
+	c.Assert(modeenv.WriteTo(""), IsNil)
+
+	var readSeedSeenLabels []string
+	restore := boot.MockSeedReadSystemEssential(func(seedDir, label string, essentialTypes []snap.Type, tm timings.Measurer) (*asserts.Model, []*seed.Snap, error) {
+		// the mock bootloader can only mock a single recovery boot
+		// chain, so pretend both seeds use the same kernel, but keep track of the labels
+		readSeedSeenLabels = append(readSeedSeenLabels, label)
+		kernelSnap := &seed.Snap{
+			Path: "/var/lib/snapd/seed/snaps/pc-kernel_1.snap",
+			SideInfo: &snap.SideInfo{
+				RealName: "pc-kernel",
+				Revision: snap.Revision{N: 1},
+			},
+		}
+		return s.uc20dev.Model(), []*seed.Snap{kernelSnap}, nil
+	})
+	defer restore()
+
+	resealCalls := 0
+	restore = boot.MockSecbootResealKeys(func(params *secboot.ResealKeysParams) error {
+		resealCalls++
+		c.Logf("params.ModelParams: %+q", params.ModelParams[0])
+		c.Logf("key files: %q", params.KeyFiles)
+
+		// bootloader variables have not been modified yet
+		c.Check(mtbl.BootVars, HasLen, 0)
+		return nil
+	})
+	defer restore()
+
+	err := boot.SetTryRecoverySystem(s.uc20dev, "1234")
+	c.Assert(err, IsNil)
+
+	c.Check(mtbl.BootVars, DeepEquals, map[string]string{
```

Comment:
let's try and use GetBootVars as much as possible so we can stop exporting BootVars at some point and make the mock bootloaders act more like real bootloaders

Reply:
We may still need to have something line GetAllBootVars() to make sure nothing unexpected was set.

Reply:
yes agreed on that